appservice/src/intTest/groovy/nu/yona/server/UserPhotoTest.groovy

```diff
@@ -0,0 +1,290 @@
+/*******************************************************************************
+ * Copyright (c) 2015 Stichting Yona Foundation
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v.2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at https://mozilla.org/MPL/2.0/.
+ *******************************************************************************/
+package nu.yona.server
+
+import org.apache.http.HttpEntity
+import org.apache.http.entity.mime.MultipartEntityBuilder
+import org.apache.http.entity.mime.content.InputStreamBody
+import org.codehaus.groovy.runtime.MethodClosure
+
+import groovy.json.*
+import nu.yona.server.test.User
+
+class UserPhotoTest extends AbstractAppServiceIntegrationTest
+{
+	def 'User uploads user photo'()
+	{
+		given:
+		def richard = addRichard()
+		appService.yonaServer.restClient.encoder.putAt("multipart/form-data", new MethodClosure(this, 'encodeMultipartEntity'))
+		def multipartEntity = MultipartEntityBuilder.create()
+				.addPart("file", new InputStreamBody(new ByteArrayInputStream(Base64.getDecoder().decode("R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7")), "image/gif", "MyPhoto.png"))
+				.build()
+
+		when:
+		def response = appService.yonaServer.restClient.put(path: richard.editUserPhotoUrl, requestContentType :"multipart/form-data", headers: ["Yona-Password": richard.password], body: multipartEntity)
+
+		then:
+		response.status == 200
+		response.contentType == "application/json"
+		def newUserPhotoUrl = response.responseData?._links?.self?.href
+		newUserPhotoUrl != null
+
+		def richardAfterUpdate = appService.reloadUser(richard, appService.&assertUserGetResponseDetailsWithPrivateDataWithUserPhoto)
+		richardAfterUpdate.userPhotoUrl == newUserPhotoUrl
+
+		cleanup:
+		appService.deleteUser(richard)
+	}
+
+	def 'Download user photo'()
+	{
+		given:
+		def richard = addRichard()
+		def userPhotoUrl = uploadUserPhoto(richard)
+
+		when:
+		def response = appService.yonaServer.restClient.get(path: userPhotoUrl)
+
+		then:
+		response.status == 200
+		response.contentType == "image/png"
+
+		cleanup:
+		appService.deleteUser(richard)
+	}
+
+	def 'User updates user photo; the previous photo is still retrievable'()
+	{
+		given:
+		def richard = addRichard()
+		def previousUserPhotoUrl = uploadUserPhoto(richard)
+
+		when:
+		def newUserPhotoUrl = uploadUserPhoto(richard)
+
+		then:
+		def richardAfterUpdate = appService.reloadUser(richard, appService.&assertUserGetResponseDetailsWithPrivateDataWithUserPhoto)
+		richardAfterUpdate.userPhotoUrl == newUserPhotoUrl
+
+		def retrievePreviousPhotoResponse = appService.yonaServer.restClient.get(path: previousUserPhotoUrl)
```

Comment:
Assert the retrievability of the new one too, and verify that both URLs are different.

Reply:
Good point, added asserts